full-stack-intro/client/src/components/Todos.tsx

```diff
@@ -0,0 +1,156 @@
+import { useEffect, useState } from 'react';
+import { PageTitle } from './PageTitle';
+import { TodoList } from './TodoList';
+import { TodoForm } from './TodoForm';
+
+export type UnsavedTodo = {
+  task: string;
+  isCompleted: boolean;
+};
+export type Todo = UnsavedTodo & {
+  todoId: number;
+};
+
+export function Todos() {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<unknown>();
+
+  /* Implement useEffect to fetch all todos. Hints are at the bottom of the file. */
+  useEffect(() => {
+    async function read() {
+      try {
+        const response = await fetch('/api/todos');
+        if (!response.ok) throw new Error(`Error: ${response.status}`);
+
+        const todoRead = await response.json();
+        setTodos(todoRead);
+      } catch (err) {
+        setError(err);
+      } finally {
+        setIsLoading(false);
+      }
+    }
+
+    read();
+  }, []);
+
+  /* Implement addTodo to add a new todo. Hints are at the bottom of the file. */
+  async function addTodo(newTodo: UnsavedTodo) {
+    try {
+      const response = await fetch('/api/todos', {
+        method: 'POST',
+        body: JSON.stringify(newTodo),
+        headers: { 'Content-Type': 'application/json' },
+      });
+      if (!response.ok) throw new Error(`Error: ${response.status}`);
+      const addedTodo = await response.json();
+      const updateTodos = structuredClone(todos);
+      updateTodos.push(addedTodo);
+      setTodos(updateTodos);
+    } catch (err) {
+      setError(err);
+    }
+  }
+
+  /* Implement toggleCompleted to toggle the completed state of a todo. Hints are at the bottom of the file. */
+  async function toggleCompleted(todo: Todo) {
+    try {
+      const response = await fetch(`/api/todos/${todo.todoId}`, {
+        method: 'PUT',
+        body: JSON.stringify({
+          task: todo.task,
+          isCompleted: !todo.isCompleted,
+        }),
+        headers: { 'Content-Type': 'application/json' },
+      });
+      if (!response.ok) throw new Error(`Error: ${response.status}`);
+      const toggled = await response.json();
+      const updateTodos = structuredClone(todos);
+      const replaceIndex = updateTodos.findIndex(
+        (element) => element.todoId === todo.todoId
+      );
+      updateTodos.splice(replaceIndex, 1, toggled);
```

Comment:
This code is good, but since you have to clone, then find, then splice in the changed element, the code I use seems simpler:
`const updateTodos = todos.map((t) => t.todoId === todo.todoId ? todo : t);`